File: regular_practice/20220922/_A.java
import java.io.*;
public class _A {
    static class Image {
        int w, h;
        
        Image(int w, int h) {
            this.w = w;
            this.h = h;
        }
    }

    static class Line {
        int w, h;

        Line() { this(0, 0); }

        Line(int w, int h) {
            this.w = w;
            this.h = h;
        }

        void add(Image img) {
            int imgWidth = img.w, imgHeight = img.h;
            if (w + imgWidth > M) {
                imgHeight = (imgHeight * (M - w) + imgWidth - 1) / imgWidth;
                imgWidth = M - w;
            }
            this.w += imgWidth;
            this.h = Math.max(this.h, imgHeight);
        }

        boolean isFull() {
            return this.w == M;
        }

        boolean isEmpty() {
            return this.w == 0 && this.h == 0;
        }

        void clear() {
            this.w = 0;
            this.h = 0;
        }

        Line copy() {
            return new Line(this.w, this.h);
        }
    }

    static int M, N;
    static Image[] imgs;
    static int[] f;

    public static void main(String[] args) throws IOException {
        StreamTokenizer in = new StreamTokenizer(new BufferedReader(new InputStreamReader(System.in)));
        in.nextToken();
        M = (int) in.nval;
        in.nextToken();
        N = (int) in.nval;
        imgs = new Image[N];
        f = new int[N];
        for (int i = 0; i < N; ++i) {
            in.nextToken();
            int w = (int) in.nval;
            in.nextToken();
            int h = (int) in.nval;
            imgs[i] = new Image(w, h);
        }
        for (int i = N - 1; i >= 0; --i) {
            f[i] = addTillFull(new Line(), i);
        }
        /*
         * 
         */
        Line line = new Line();
        int minHeight = Integer.MAX_VALUE, totalHeight = 0;
        for (int del = 0; del < N; ++del) {
            int newHeight = totalHeight + (line.isEmpty() && del + 1 < N ?
                f[del + 1] : addTillFull(line.copy(), del + 1));
            minHeight = Math.min(minHeight, newHeight);
            line.add(imgs[del]);
            if (line.isFull()) {
                totalHeight += line.h;
                line.clear();
            }
        }
        System.out.println(minHeight);
    }

    static int addTillFull(Line line, int k) {
        for (; k < N && !line.isFull(); ++k) {
            line.add(imgs[k]);
        }
        return k < N ? line.h + f[k] : line.h;
    }
}

File: regular_practice/20220922/_A.cpp
#include <iostream>

using namespace std;

const int MAX_N = 1e5 + 10;
int N, M;
int f[MAX_N];

struct Image {
    int w, h;
} imgs[MAX_N];

struct Line {
    int w, h;

    Line(int w = 0, int h = 0): w(w), h(h) { }

    Line operator+(const Image& img) {
        int img_w = img.w, img_h = img.h;
        if (w + img_w > M) {
            img_h = (img_h * (M - w) + img_w - 1) / img_w;
            img_w = M - w;
        }
        return Line(w + img_w, max(h, img_h));
    }

    Line operator+=(const Image& img) {
        return *this = *this + img;
    }

    bool full() {
        return w == M;
    }

    bool empty() {
        return w == 0 && h == 0;
    }

    void clear() {
        w = 0;
        h = 0;
    }
};

int addTillFull(Line line, int k) {
    for (; !line.full() && k < N; ++k) line += imgs[k];
    return line.h + f[k];
}

int main() {
    scanf("%d%d", &M, &N);
    for (int i = 0; i < N; ++i) scanf("%d%d", &imgs[i].w, &imgs[i].h);
    for (int i = N - 1; i >= 0; --i) f[i] = addTillFull(Line(), i);
    Line line; 
    int min_height = 0x3f3f3f3f, total_height = 0;
    for (int del = 0; del < N; ++del) {
        int new_height = total_height + (line.empty() ?
            f[del + 1] : addTillFull(line, del + 1));
        min_height = min(min_height, new_height);
        line += imgs[del];
        if (line.full()) {
            total_height += line.h;
            line.clear();
        }
    }
    printf("%d\n", min_height);

    return 0;
}